lib/semian/lru_hash.rb

```diff
@@ -0,0 +1,63 @@
+class LRUHash
+  extend Forwardable
+  def_delegators :@table, :size, :count, :empty?, :keys, :clear, :delete
+  attr_reader :table, :minimum_time_in_lru
+
+  class NoopMutex
+    def synchronize(*)
+      yield
+    end
+  end
+
+  def initialize(minimum_time_in_lru:)
+    @minimum_time_in_lru = minimum_time_in_lru
+    @table = {}
+    @lock =
+      if Semian.thread_safe?
+        Mutex.new
+      else
+        NoopMutex.new
+      end
+  end
+
+  def set(key, value)
+    @lock.synchronize do
+      delete(key)
+      @table[key] = value
+    end
+    clear_unused_resources
+    value
+  end
+
+  def get(key)
+    @lock.synchronize do
+      if found = delete(key)
+        @table[key] = found
+      end
+      @table[key]
+    end
+  end
+
+  def []=(key, value)
+    set(key, value)
+  end
+
+  def [](key)
+    get(key)
+  end
+
+  private
+
+  def clear_unused_resources
+    @table.each do |key, value|
```

Comment:
We're doing an O(n) operation here which means there's absolutely no value in this being an LRU to begin with.

Maybe we should parameterize how many elements of the table are looked at.

As we discussed over lunch of a few days ago: `clear_unused_resources(2)` should do the trick.

Another optimization I'm pretty sure we can do here (though it may not be necessary) is that once we skip a resource because it's in use, as can assume the future items in the queue will also be in use. (Assuming you implement `updated_at` rather than `created_at`.)